### ISSUE TITLE
Perhaps forked repos can have issues too.

### ISSUE THREAD

**Issue by mehmetf** [mehmetf/flutter#1]
< None >

**Comment by mehmetf** [comment]
@aletorrado I am just trying to move clarification discussion here from https://github.com/flutter/flutter/issues/6827 which is already overinflated with many  comments.

Let's see if this works.

Could you please clarify exactly what you did and how you observed Google Ads app saving state. I am particularly interested in how you knew the OS was killing the process: https://github.com/flutter/flutter/issues/6827#issuecomment-447488177 .

**Comment by mehmetf** [comment]
@gitspeak @Zhuinden as well

@gitspeak We got off the wrong foot. I deleted my comments and yours because we are unnecessarily crowding the discussion.

I was referring to the very first post of that bug ([more than 2 years ago](https://github.com/flutter/flutter/issues/6827#issue-188885146)) where the author used "Dont keep activities" technique to test the Flutter app saving state. I assumed that's what @aletorrado was doing as well. I asked him above to clarify how he knew how Google Ads was killed by the OS.

One thing is for sure: Google Ads is not using `onSaveInstanceState`. I am sure of it because I am a contributor to their code. Let's put our heads together and figure out what @aletorrado saw and how we can explain it.

**Comment by mehmetf** [comment]
@gamebox This is a test for notifications.

**Comment by Zhuinden** [comment]
The only way to know if the process *came back from death* is if you have a static boolean that is set to true by the first Activity to launch in onCreate, but this boolean is false and the savedInstanceState is non-null.

The only way to guarantee that you saved things when you're elligible for termination is if you do it in onStop (or your navigation is transactional and persisted to disk).

If the activity savedInstanceState bundle is null, then the app has a clean start, which means the persisted nav state should be discarded.

For this to work properly, the launchMode should be singleTask

**Comment by mehmetf** [comment]
@Zhuinden I wasn't asking how to know whether the process came back from death or not. I am trying to understand what exactly @aletorrado saw when he informed @gitspeak that Google Ads is able to preserve the state when killed.  I sincerely think @aletorrado enabled "Don't keep activities" to test Google Ads. I just tested a bunch of apps with my phone and that's the only way I can reliably get the activity to be destroyed.

Swiping the app from the task switcher or force stopping the app kills the process which, as both of you pointed out, we don't expect apps to store state but may be they still can? I did some experiments and observed that onSaveInstanceState is called as soon as app is put into background including invoking the task switcher. So, in theory, you could save the state prior to app being force killed.

I would assume onSaveInstanceState is called on the UI thread. I also did some cursory research into any data size limits. According to https://developer.android.com/guide/components/activities/parcelables-and-bundles, there's a cross-process limit of about 1M. That might rule out serializing the isolate in its entirety. This needs further experimentation.

I like @gitspeak's [response](https://github.com/flutter/flutter/issues/6827#issuecomment-448252611) to the proposed solution. I also think it should not be left to developer's to save the app's state. 

There are two problems here, that I can see:
1. Determining what to save
2. Saving it in a synchronous fashion via onSaveInstanceState w/o running into race conditions with Android's UI thread

(2) tells me that we might need to operate directly on the isolate w/o having to reach into the Dart layer. (1) tells me we should probably be able to split the isolate memory into multiple sections; assets, data and code. Perhaps we can simply save the data. Assuming that code should be exactly the same (and assets are loaded lazily), serializing the data aspect should suffice. This is akin to saving the "State Tree" that @gitspeak was alluding to.

I would assume when code changes (app is updated for instance), instance state is reset by the OS so we should never run into a case where some Flutter code is called with an out-of-date data bundle.

What do you guys think?

**Comment by gitspeak** [comment]
@mehmetf 

>I did some experiments and observed that onSaveInstanceState is called as soon as app is put into background including invoking the task switcher. So, in theory, you could save the state prior to app being force killed.

There's nothing special about the task switcher (it's a regular Android app). Each time an Android App is sent to the background it's eligible for process termination (unless an App Service was activated before hand). Pressing the "Home Button" is one (quick) way to do it. Before an app is actually sent to background two critical things are *guaranteed to happen* 1) onSaveInstanceState is invoked 2) onStop is invoked and *in that order* (since Android 3.0). The motivation for that is to have the App process in a state where it can be safely terminated.

> I would assume onSaveInstanceState is called on the UI thread. 

Correct. 

>I would assume when code changes (app is updated for instance), instance state is reset by the OS so we should never run into a case where some Flutter code is called with an out-of-date data bundle

Correct. Android will brutally terminate the process without saving any state.

**Comment by gitspeak** [comment]
>So, in theory, you could save the state prior to app being force killed.

I figure you mean “Force Stop” ? (The “Application Manager” Button feature next to Uninstall)

There is always an opportunity for saving state when an app is sent to the background; However, in this case pressing the “Force Stop” button will, by design, Terminate the process and discard any saved state like when you update the app or discard it from the task switcher, since all these cases are user-driven and it makes sense to behave this way.

**Comment by aletorrado** [comment]
> @Zhuinden I wasn't asking how to know whether the process came back from death or not. I am trying to understand what exactly @aletorrado saw when he informed @gitspeak that Google Ads is able to preserve the state when killed. I sincerely think @aletorrado enabled "Don't keep activities" to test Google Ads. I just tested a bunch of apps with my phone and that's the only way I can reliably get the activity to be destroyed.

@mehmetf that's correct. I've enabled the "Don't keep activities" switch and tested that way. Don't know if it implies killing the process or not.

**Comment by mehmetf** [comment]
Thanks for clarifying :+1: 

To be honest, I was surprised that Don't keep activities did not kill the process but yes, since then, I confirmed (by `adb shell ps`) that the process lives with that option turned on which explains why Google Ads "keeps its state".

**Comment by Zhuinden** [comment]
No, don't keep activities only happens if that is enabled in dev options. It never happens in "real life"

**Comment by gitspeak** [comment]
Activities and Processes are orthogonal concepts. "Killing" / "Destroying" an Activity merely means releasing a Java object graph to be garaged collected. It says nothing about the Java Process itself or any other Android component within the Process.

**Comment by gitspeak** [comment]
.. and please don't mix terms like: "OS Kills the Activity" the OS, linux, knows about Processes and how to kill them if asked by the "Android Runtime". The OS doesn't no anything about Activities.

**Comment by mehmetf** [comment]
Before it gets lost in discussion, let me ask my question again as it is important in terms of pushing the priority of the issue we are discussing about: "Other than user-driven use cases, memory pressure and being backgrounded for a long time, did you observe Android runtime asking the OS to kill an app process?"

**Comment by Zhuinden** [comment]
"Memory pressure" yes, "long time" is random. If I open the Camera app and Skype, then the 3rd app I came from tends to die. :thinking: 

People often say "oh but low memory happens only in 6 hours" but I think it's more-so 2-4 minutes.

I've seen apps restart after getting a 10 minute phone call.

**Comment by gitspeak** [comment]
@mehmetf 

>Makes sense. Other than user-driven use cases, memory pressure and being backgrounded for a long time, did you observe Android killing an app process?

I mentioned previously (deleted post) that in my experience Android will terminate a background process even when there is no memory pressure at all.. and that the actual decision making is rooted in heuristics so I suggest you don't get too attached to particular use-cases but rather study the Android life-cycle semantics carefully and ensure Flutter on Android does the same (or at least avoid doing anything against :))


>If these are the main reasons, I would think this is far more important for the emerging markets like India where low-end phones are more common. That could be one way to motivate a solution for this problem.

Sure! However, in IMHO a bigger motivation would be that you are now establishing a core Flutter programming contract that will enable it to work  on *Any Android Device - Everywhere*.

**Comment by gitspeak** [comment]
>I've seen apps restart after getting a 10 minute phone call.

yea, that happens quite often :) and can easily be observed if the browser was the active app at time the call came in, you notice the page reload after the Dialer app is dismissed.

**Comment by gitspeak** [comment]
@mehmetf

>Before it gets lost in discussion, let me ask my question again as it is  important in terms of pushing the priority of the issue we are  discussing about: "Other than user-driven use cases, memory pressure and  being backgrounded for a long time, did you observe Android runtime  asking the OS to kill an app process?

If you must, here is one more "juicy" use case: Consider an Android App that includes a Service Component and no Activity Components whatsoever. The App is configured to run at device boot. If that App Process runs too long the Android Runtime will notice that and request the OS to kill it. Obviously, the key question here is how long is "too long" and that, as I mentioned, is determined by  heuristics which are internal to the runtime and you *definitely* don't want to take any dependency on that since you have no guarantee whatsoever that it's consistent across Android versions and I'm betting it ain't.

**Comment by hey99xx** [comment]
> "Other than user-driven use cases, memory pressure and being backgrounded for a long time, did you observe Android runtime asking the OS to kill an app process?"

Generally 10+ minutes I have observed. While @Zhuinden is absolutely right that "Don't keep activities" does not happen in real life, this does in happen in real life. To get around periods of waiting during development, Android does offer you a convenience though. You can go to your device's developer option, and limit background process count. This should be right below "Don't keep activities" setting. Set it to 2 -> open your app -> go to Chrome -> go to Camera -> open your app again from recents and you will emulate instance state restoring.
